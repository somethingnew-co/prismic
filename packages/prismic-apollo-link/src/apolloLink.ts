import Prismic from 'prismic-javascript'
import { ApolloLink, HttpLink } from '@apollo/client'
import { setContext } from '@apollo/client/link/context'

interface Config {
  repositoryName: string
  accessToken?: string
  defaultLocale?: string
}

interface PreviewData {
  [prop: string]: any
}

interface PrismicApolloLink {
  (config: Config, previewData?: PreviewData): ApolloLink
}

const prismicApolloLink: PrismicApolloLink = (config, previewData) => {
  const { repositoryName, accessToken, defaultLocale } = config
  const apiEndpoint = `https://${repositoryName}.cdn.prismic.io/api/v2`
  const graphqlEndpoint = `https://${repositoryName}.prismic.io/graphql`

  const prismicLink = setContext(async (req, previousContext) => {
    const prismicClient = Prismic.client(apiEndpoint, { accessToken })
    const prismicApi = await prismicClient.getApi()

    const headers = {
      ...previousContext.headers,
      'Accept-Language': defaultLocale || 'en-us',
      'Prismic-ref': previewData?.ref || prismicApi.masterRef.ref,
    }

    if (accessToken) {
      headers.Authorization = `Token ${accessToken}`
    }

    return { headers }
  })

  const httpLink = new HttpLink({
    fetch: (url, options) => fetch(url, options),
    uri: graphqlEndpoint,
    useGETForQueries: true,
  })

  return ApolloLink.from([prismicLink, httpLink])
}

export default prismicApolloLink
