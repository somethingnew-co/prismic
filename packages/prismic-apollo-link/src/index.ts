export { default as prismicApolloLink } from './apolloLink'
