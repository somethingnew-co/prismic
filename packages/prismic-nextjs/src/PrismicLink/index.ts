export { default as PrismicLink } from './PrismicLink'
