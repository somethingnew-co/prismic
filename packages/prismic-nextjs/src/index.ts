export { Link } from './Link'
export * from '@stnew/prismic-react'
