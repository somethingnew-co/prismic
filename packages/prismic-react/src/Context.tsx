import { createContext } from 'react'
import { Context } from './types'

const prismicContext: Context = {
  slices: {},
  linkResolver: () => '/',
  hrefResolver: () => '/',
  rootResolver: '',
}

const PrismicContext = createContext(prismicContext)

export default PrismicContext
