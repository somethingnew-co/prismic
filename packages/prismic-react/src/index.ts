export { default as PrismicContext } from './Context'
export { default as PrismicProvider } from './Provider'
export { default as usePrismic } from './usePrismic'
export { default as useHtmlSerializer } from './useHtmlSerializer'
export { SliceZone } from './SliceZone'
export { default as serializeElements } from './serializeElements'
export * from '@stnew/prismic'
