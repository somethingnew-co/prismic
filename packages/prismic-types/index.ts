export interface PrismicLink {
  link_type: 'Document' | 'Web' | 'Media' | 'Any'
  url?: string
  id?: string
  lang?: string
  slug?: string
  tags?: string[]
  type?: string
  uid?: string
  isBroken?: boolean
  target?: string
  data?: any
}

export type PrismicDoc = PrismicLink

export interface PrismicSlice {
  slice_type: string
  slice_label: string | null
  primary: any
  items: any[]
}

export type PrismicDropdown = string

export type PrismicKeyText = string

export interface PrismicDocument {
  [key: string]: any
  body: PrismicSlice[]
}

export interface PrismicText {
  type: string
  text: string
  spans: any[]
}

export interface LinkResolver {
  (doc: PrismicLink): string
}

export interface Route {
  href: string
  page?: string
  root?: string
}

export interface Routes {
  [key: string]: Route
}

export interface Resolver {
  (routes: Routes): LinkResolver
}
