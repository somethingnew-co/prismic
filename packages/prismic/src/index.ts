export { default as urlResolver } from './urlResolver'
