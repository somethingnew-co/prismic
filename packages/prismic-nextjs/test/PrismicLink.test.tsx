import React from 'react'
import { PrismicLink } from '../src'
import { renderWithPrismicProvider } from './test-setup'
import { PrismicDoc } from '@stnew/prismic-types'
import { docs } from '../../../test/data'

const media: PrismicDoc = {
  link_type: 'Media',
  url: 'https://somethingnew.co/share.png',
}

const web: PrismicDoc = {
  link_type: 'Web',
  url: 'https://somethingnew.co',
}

const webEx: PrismicDoc = {
  link_type: 'Web',
  url: 'https://somethingnew.co',
  target: '_blank',
}

const testQuery = { query: { test_query: 'test query ! 123' } }

test('Prismic Doc', () => {
  for (let i = 0; i < docs.length; i += 1) {
    const { container } = renderWithPrismicProvider(
      <PrismicLink link={docs[i]}>Document link</PrismicLink>,
    )
    expect(container).toMatchSnapshot()
  }
})

test('Prismic Doc with Query', () => {
  for (let i = 0; i < docs.length; i += 1) {
    const { container } = renderWithPrismicProvider(
      <PrismicLink link={docs[i]} urlObject={testQuery}>Document link</PrismicLink>,
    )
    expect(container).toMatchSnapshot()
  }
})

test('Other Links', () => {
  const { container } = renderWithPrismicProvider(<>
    <PrismicLink link={media}>Media link</PrismicLink>
    <PrismicLink link={web}>Web link</PrismicLink>
    <PrismicLink link={webEx}>Web link (target blank)</PrismicLink>
  </>)

  expect(container).toMatchSnapshot()
})
