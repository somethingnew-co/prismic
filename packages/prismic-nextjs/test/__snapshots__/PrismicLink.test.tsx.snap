// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Other Links 1`] = `
<div>
  <a
    href="https://somethingnew.co/share.png"
    rel="noopener noreferrer"
    target="_blank"
  >
    Media link
  </a>
  <a
    href="https://somethingnew.co"
  >
    Web link
  </a>
  <a
    href="https://somethingnew.co"
    rel="noopener noreferrer"
    target="_blank"
  >
    Web link (target blank)
  </a>
</div>
`;

exports[`Prismic Doc 1`] = `
<div>
  <a
    href="/"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 2`] = `
<div>
  <a
    href="/services"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 3`] = `
<div>
  <a
    href="/developers"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 4`] = `
<div>
  <a
    href="/company"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 5`] = `
<div>
  <a
    href="/pricing"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 6`] = `
<div>
  <a
    href="/teams"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 7`] = `
<div>
  <a
    href="/teams/blue-team"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 8`] = `
<div>
  <a
    href="/teams/red-team"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 9`] = `
<div>
  <a
    href="/blog"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 10`] = `
<div>
  <a
    href="/blog/hello-world"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 11`] = `
<div>
  <a
    href="/blog/hola-mundo"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 12`] = `
<div>
  <a
    href="/privacy"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc 13`] = `
<div>
  <a
    href="/privacy/terms-and-conditions"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 1`] = `
<div>
  <a
    href="/?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 2`] = `
<div>
  <a
    href="/services?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 3`] = `
<div>
  <a
    href="/developers?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 4`] = `
<div>
  <a
    href="/company?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 5`] = `
<div>
  <a
    href="/pricing?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 6`] = `
<div>
  <a
    href="/teams?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 7`] = `
<div>
  <a
    href="/teams/blue-team?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 8`] = `
<div>
  <a
    href="/teams/red-team?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 9`] = `
<div>
  <a
    href="/blog?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 10`] = `
<div>
  <a
    href="/blog/hello-world?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 11`] = `
<div>
  <a
    href="/blog/hola-mundo?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 12`] = `
<div>
  <a
    href="/privacy?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;

exports[`Prismic Doc with Query 13`] = `
<div>
  <a
    href="/privacy/terms-and-conditions?test_query=test+query+%21+123"
  >
    Document link
  </a>
</div>
`;
